### PR TITLE
[ENH] - Add Prerender IModuleControl property (similar to RenderMode)…

### PR DESCRIPTION
**Pull request by vnetonline**
… #4178

@sbwalker I had a go at adding Prerender just like RenderMode, I hope my implementation meets your standards let me know if you would like me to modify anything

### CODE REVIEW

**Comment by sbwalker** [comment]
@vnetonline this looks good... the only question I have is related to the default behavior:

Currently there is a Site setting for Prerender which controls whether or not pre-rendering is enabled for the entire site. Currently it is possible to set Prerendering to False at the Site level and only enable Prerendering for some specific components (ie. only some Interactive components on public facing pages which need to be indexed for SEO purposes). Based on this PR, the scenario I described would not be possible due to the default value and logic below:

ModuleState.Prerender != false ? PageState.Site.Prerender : ModuleState.Prerender

Basically this says that if the component-level property is set to True then it should use the Site Setting - Prerender property setting (which would be False in the case I described above). Perhaps a more appropriate solution would be if the component-level Prerender property was a nullable boolean type and had a default value of null. Then the logic  would be able to determine if the value was overridden or not:

ModuleState.Prerender == null ? PageState.Site.Prerender : ModuleState.Prerender

The above logic uses the Site Setting unless it is explicitly overridden at the component level.

Thoughts?

**Comment by vnetonline** [comment]
@sbwalker, the above logic sounds good, I would reverse it from a conventional perspective, it's more readable like this 

`ModuleState.Prerender != null ?  ModuleState.Prerender : PageState.Site.Prerender`

I will modify the PR and submit it againt

**Comment by vnetonline** [comment]
@sbwalker i can do what you suggested using below code block ... but it doesn't look elegant enough 

```
if (ModuleState.Prerender != false)
{
    <RenderModeBoundary ModuleState="@ModuleState" PageState="@PageState" SiteState="@SiteState" 
        @rendermode="@InteractiveRenderMode.GetInteractiveRenderMode(PageState.Site.Runtime, false)" />

} 
else if (ModuleState.Prerender == true)
{
    <RenderModeBoundary ModuleState="@ModuleState" PageState="@PageState" SiteState="@SiteState" 
        @rendermode="@InteractiveRenderMode.GetInteractiveRenderMode(PageState.Site.Runtime, true)" />
}
else
{
    <RenderModeBoundary ModuleState="@ModuleState" PageState="@PageState" SiteState="@SiteState" 
        @rendermode="@InteractiveRenderMode.GetInteractiveRenderMode(PageState.Site.Runtime, 
        PageState.Site.Prerender)"/>
}
```

**Comment by vnetonline** [comment]
> @sbwalker i can do what you suggested using below code block ... but it doesn't look elegant enough
> 
> ```
> if (ModuleState.Prerender == false)
> {
>     <RenderModeBoundary ModuleState="@ModuleState" PageState="@PageState" SiteState="@SiteState" 
>         @rendermode="@InteractiveRenderMode.GetInteractiveRenderMode(PageState.Site.Runtime, false)" />
> 
> } 
> else if (ModuleState.Prerender == true)
> {
>     <RenderModeBoundary ModuleState="@ModuleState" PageState="@PageState" SiteState="@SiteState" 
>         @rendermode="@InteractiveRenderMode.GetInteractiveRenderMode(PageState.Site.Runtime, true)" />
> }
> else
> {
>     <RenderModeBoundary ModuleState="@ModuleState" PageState="@PageState" SiteState="@SiteState" 
>         @rendermode="@InteractiveRenderMode.GetInteractiveRenderMode(PageState.Site.Runtime, 
>         PageState.Site.Prerender)"/>
> }
> ```

Simplefied Version 

```
    var rendermode = InteractiveRenderMode.GetInteractiveRenderMode(PageState.Site.Runtime, ModuleState.Prerender ? 
       PageState.Site.Prerender);

    <RenderModeBoundary ModuleState="@ModuleState" PageState="@PageState" SiteState="@SiteState" 
        @rendermode="rendermode" />
```

**Comment by sbwalker** [comment]
@vnetonline the PR is already merged

**Comment by vnetonline** [comment]
> @vnetonline the PR is already merged

Yes thank you was just adding a comment as to what changed in this commit
